eos-migrate-chromium-profile

```diff
@@ -0,0 +1,151 @@
+    """Update file associations, which in particular includes x-scheme-handler/http and
+    friends to specify the default web browser.
+
+    We cannot use GLib's own API to query what mime types the old Chromium desktop file
+    is a handler for, because we can't construct a GDesktopAppInfo for it, because its
+    desktop file no longer exists.
+    """
+
+    keyfile = GLib.KeyFile()
+    try:
+        keyfile.load_from_file(
+            path, GLib.KeyFileFlags.KEEP_COMMENTS | GLib.KeyFileFlags.KEEP_TRANSLATIONS,
+        )
+    except GLib.GError as gerror:
+        if gerror.matches(GLib.file_error_quark(), GLib.FileError.NOENT):
+            return
+
+        raise
+
+    changed = False
+
+    for group in MIMEAPPS_GROUPS:
+        if not keyfile.has_group(group):
+            continue
+
+        keys, _length = keyfile.get_keys(group)
+        for key in keys:
+            values = keyfile.get_string_list(group, key)
+            try:
+                i = values.index(OLD_DESKTOP_FILE)
+            except ValueError:
+                pass
+            else:
+                values[i] = NEW_DESKTOP_FILE
+                keyfile.set_string_list(group, key, values)
+                changed = True
+
+    if changed:
+        keyfile.save_to_file(path)
+
+
+def update_desktop_shortcut(path):
+    keyfile = GLib.KeyFile()
+    print("path: " + path)
+    keyfile.load_from_file(path, GLib.KeyFileFlags.NONE)
+
+    group = "Desktop Entry"
+    if not keyfile.has_group(group):
+        return
+
+    exec_cmd = keyfile.get_string(group, "Exec")
+    if "/usr/bin/chromium-browser" not in exec_cmd:
+        return
+
+    keyfile.set_string(group, "Exec",
+                       exec_cmd.replace("/usr/bin/chromium-browser",
+                                        "/usr/bin/flatpak run org.chromium.Chromium"))
+    keyfile.set_string(group, "X-Flatpak-Part-Of", "org.chromium.Chromium")
+    keyfile.save_to_file(path)
+
+
+def update_desktop_shortcuts():
+    for filename in os.listdir(DESKTOP_SHORTCUTS_DIR):
+        if filename.endswith(".desktop"):
+            update_desktop_shortcut(os.path.join(DESKTOP_SHORTCUTS_DIR,
+                                                 filename))
+
+
+def update_old_config_references(path):
+    if not os.path.isfile(path):
+        return
+
+    for line in fileinput.input(path, inplace=True):
+        line = line.replace(OLD_CHROMIUM_CONFIG_DIR, NEW_CHROMIUM_CONFIG_DIR)
+        line = line.replace(OLD_WIDEVINE_SYSTEM_DIR, "")
+        print(line, end='')
+
+
+def main():
+    migrated_file = os.path.join(NEW_CHROMIUM_DATA_DIR, ".migrated")
+    if os.path.exists(migrated_file):
+        return
+
+    update_mimeapps_list(MIMEAPPS_LIST)
+    update_desktop_shortcuts()
+
+    if (
+        os.path.isdir(OLD_CHROMIUM_CONFIG_DIR) and
+        not os.path.isdir(NEW_CHROMIUM_CONFIG_DIR)
+    ):
+        os.makedirs(NEW_CHROMIUM_CONFIG_DIR)
+        os.rename(OLD_CHROMIUM_CONFIG_DIR, NEW_CHROMIUM_CONFIG_DIR)
```

Comment:
TIL that you can rename over a directory that exists. I would normally spell this as:

```python3
        os.makedirs(os.path.dirname(NEW_CHROMIUM_CONFIG_DIR), exist_ok=True)
        os.rename(OLD_CHROMIUM_CONFIG_DIR, NEW_CHROMIUM_CONFIG_DIR)
```

but it does appear that what you have works.

Reply:
My idea here is that we wouldn't migrate the user profile if the user had already ran the flatpak (`NEW_CHROMIUM_CONFIG_DIR` already exists). In that case we would only update the mime associations and desktop shortcuts.

Left as is but can update if you prefer.

Reply:
Yes, I understand the intent of the `not os.path.isdir()` check. I was more surprised that creating an empty directory then renaming over it works.

Reply:
Ah, I misread your comment and didn't notice the `os.path.dirname`, fixed it now.